test/TestUserAcceptance.hs

```diff
@@ -42,198 +49,138 @@ testTimeout = 10 ^ 6 * 8
+
+          liftIO $ step "apply"
+          out <- sendKeys "Enter" "Item 0 of 1"
+          assertSubstrInOutput "Item 0 of 1" out
+
+          liftIO $ step "view currently selected mail"
+          out <- sendKeys "Enter" "HOLY PUREBRED"
+          assertSubstrInOutput "HOLY PUREBRED" out
 
 testUserCanSwitchBackToIndex ::
   TestTree
 testUserCanSwitchBackToIndex =
-    tmuxSession "user can switch back to mail index during composition" steps
-  where
-    steps =
-        [ ApplicationStep
-              "m"
-              "start composition"
-              False
-              "From"
-              assertSubstrInOutput
-        , ApplicationStep
-              "testuser@foo.test\r"
-              "enter from email"
-              False
-              "To"
-              assertSubstrInOutput
-        , ApplicationStep
-              "user@to.test\r"
-              "enter to: email"
-              False
-              "Subject"
-              assertSubstrInOutput
-        , ApplicationStep
-              "test subject\r"
-              "enter subject"
-              False
-              "~"
-              assertSubstrInOutput
-        , ApplicationStep
-              "iThis is a test body"
-              "enter mail body"
-              False
-              "body"
-              assertSubstrInOutput
-        , ApplicationStep
-              "Escape"
-              "exit insert mode in vim"
-              False
-              "body"
-              assertSubstrInOutput
-        , ApplicationStep
-              ": x\r"
-              "exit vim"
-              False
-              "Attachments"
-              assertSubstrInOutput
-        , ApplicationStep
-              "Tab"
-              "switch back to index"
-              False
-              "Testmail"
-              assertSubstrInOutput
-        , ApplicationStep
-              "Tab"
-              "switch back to the compose editor"
-              False
-              "test subject"
-              assertSubstrInOutput]
-
-data ApplicationStep = ApplicationStep
-    { asKeys :: String  -- ^ the actual commands to send
-    , asDescription :: String  -- ^ step definition
-    , asAsLiteralKey :: Bool  -- ^ disables key name lookup and sends literal input
-    , asExpected :: String  -- ^ wait until the terminal shows the expected string or timeout
-    , asAssertInOutput :: String -> String -> Assertion  -- ^ assert this against the snapshot
-    }
-
-assertSubstrInOutput :: String -> String -> Assertion
-assertSubstrInOutput out substr = assertBool "in out" $ substr `isInfixOf` out
+  withTmuxSession "user can switch back to mail index during composition" $
+        \step -> do
+            liftIO $ step "start composition"
+            out <- sendKeys "m" "From"
+            assertSubstrInOutput "From" out
+
+            liftIO $ step "enter from email"
+            out <- sendKeys "testuser@foo.test\r" "To"
+            assertSubstrInOutput "To" out
```

Comment:
Once refactored, I've actually realised that all these assertions are kind of superfluous since the `sendKeys` function already waits for the string I'm asserting against to appear. Perhaps an improvement another day.

Reply:
Yeah whatevs mate, I got no idea; tests are your department :)